### PR TITLE
[New Feature] Zenless Zone Zero Game Setings

### PR DESCRIPTION
**Pull request by bagusnl**
# Main Goal
Implements Zenless Game Settings to Collapse

image

image

## PR Status :
- Overall Status : Done
- Commits : Done
- Synced to base (Collapse:main) : Yes
- Build status : OK
- Crashing : No
- Bug found caused by PR : 
  - [Low] Save output of GSP for GENERAL_DATA is not indented like how the game outputs it
  - [Low] "Use Resizable Window" gets enabled when toggling "Use Borderless Windowed Mode" on and off

### CODE REVIEW

**Comment by bagusnl** [comment]
forgor how mad qodana will be on in progress feature. will reopen when stuff is mostly done

**Comment by bagusnl** [comment]
Reopening as frontend is "done"

File: CollapseLauncher/Classes/Helper/Metadata/PresetConfig.cs [hunk 1]
```diff
@@ -468,9 +468,17 @@
             int verInt = GameDataVersion.GetBytesToIntVersion(gameVersion);
             if (!value.DataVersion.TryGetValue(verInt, out GameDataVersion? verData))
             {
-                return null;
+                // Fallback to find the default value anyway
+                KeyValuePair<int, GameDataVersion>? kvpTemp = value.DataVersion.FirstOrDefault();
+                if (kvpTemp == null)
+                    return null;
+
+                verData = kvpTemp?.Value;
```
[the remark this answers is not in the document]
Comment (bagusnl) [reply CollapseLauncher/Classes/Helper/Metadata/PresetConfig.cs:478]: Surpressed

**Review by neon-nyan — APPROVED** [review]
LGTM! I Approb!

File: CollapseLauncher/XAMLs/MainApp/Pages/GameSettingsPages/ZenlessGameSettingsPage.xaml.cs [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+            
+            acH.RemoveAll(h => h > ScreenProp.GetMaxHeight());
+            
+            List<string> resPairs = new List<string>();
+
+            foreach (var h in acH)
+            {
+                int w = (int)(h * nativeAspRatio);
+                // TODO: remove identifier
+                resPairs.Add($"{w}x{h} Fullscreen");
```
Comment (shatyuka) [inline CollapseLauncher/XAMLs/MainApp/Pages/GameSettingsPages/ZenlessGameSettingsPage.xaml.cs:209]: Localize this string?

File: CollapseLauncher/XAMLs/MainApp/Pages/GameSettingsPages/ZenlessGameSettingsPage.xaml.cs [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+
+            // If res is 21:9 then add proper native to the list
+            if (Math.Abs(nativeAspRatio - ulWideRatio) < 0.01)
+                resPairs.Add($"{displayProp.Width}x{displayProp.Height}");
+            
+            foreach (var h in acH)
+            {
+                int w = (int)(h * wideRatio);
+                // TODO: remove identifier
+                resPairs.Add($"{w}x{h} Windowed");
```
Comment (shatyuka) [inline CollapseLauncher/XAMLs/MainApp/Pages/GameSettingsPages/ZenlessGameSettingsPage.xaml.cs:235]: Also this one.

Comment (neon-nyan) [reply CollapseLauncher/XAMLs/MainApp/Pages/GameSettingsPages/ZenlessGameSettingsPage.xaml.cs]: Fixed!

File: CollapseLauncher/XAMLs/MainApp/Pages/GameSettingsPages/ZenlessGameSettingsPage.xaml.cs (hunk 2 above)
Comment (neon-nyan) [reply CollapseLauncher/XAMLs/MainApp/Pages/GameSettingsPages/ZenlessGameSettingsPage.xaml.cs]: Fixed!

**Review by shatyuka — APPROVED** [review]

**Comment by bagusnl** [comment]
approball accepted, merging changes by merge commit

Thanks all for this painful GSP implementation~

**Comment by NSPC911** [comment]
you forgot to change the readme lol

**Pull request merged**